packages/wes_api/client/api_workflow_execution_service_log_extension.go

```diff
@@ -0,0 +1,146 @@
+/*
+ * Additional methods for actually retrieving data pointed to by URLs in WES responses. 
+ */
+
+package wes_client
+
+import (
+	_context "context"
+	_ioutil "io/ioutil"
+	_nethttp "net/http"
+	_neturl "net/url"
+	"fmt"
+	"strings"
+)
+
+// Linger please
+var (
+	_ _context.Context
+)
+
+/*
+GetRunLogData Get data linked to by GetRunLog.
+Returns the content of a URL reverenced in a GetRunLog response, if that URL is also on the WES server.
+ * @param ctx _context.Context - for authentication, logging, cancellation, deadlines, tracing, etc. Passed from http.Request or context.Background().
+ * @param runId The run ID used for the GetRunLog call
+ * @param dataUrl The URL string from the GetRunLog call
+@return string
+*/
+func (a *WorkflowExecutionServiceApiService) GetRunLogData(ctx _context.Context, runId string, dataUrl string) (string, *_nethttp.Response, error) {
```

Comment:
It seems like this is inspired by the OpenAPI-generated `api_workflow_execution_service.go`? Could we be a little less verbose if we don't think of this as an extension/part of the WES client?

I find WES to be a little under-specified in some places, but I'm no expert and am pretty much going off the [OpenAPI spec](https://ga4gh.github.io/workflow-execution-service-schemas/docs/#operation/GetRunLog). If I'm wrong about any of my interpretations of what this URL should be and what should be contained there, let me know.

If we're assuming these are just text files available at either an absolute URL or one relative to the WES endpoint, can we just [download that file](https://stackoverflow.com/questions/11692860/how-can-i-efficiently-download-a-large-file-using-go)? Do we need to bother with checking the WES-specified error codes and content types? I'm not sure we have any guarantees that the endpoint hosting the file will respond with the error codes/payloads that the WES client is expecting.

Not something we need to solve right now, but what's prompting me to consider a slightly less WES-endpoint-coupled approach is that most of our existing engines run in Batch, and we currently access their logs via Batch+Cloudwatch. Cloudwatch Logs don't have a standard URL scheme, but I suspect we'll (AGC) eventually want to extend this to allow for loading stdout/stderr from an S3 file or a Cloudwatch Log stream as well as via a URL.

For now, I'm just wondering if we can clean up some of this generated-ish code by just adding some plain-old file loading to `packages/cli/internal/pkg/wes/Client.go`

Reply:
It's not so much "inspired by" the OpenAPI-generated code as it is "a complete copy-paste of" the OpenAPI-generated code. I did that for consistency with the rest of the WES API implementation, since retrieving these logs is specified in WES (just in English that the code generator can't read). I also did it because I thought it would work better, faster, than me trying to write something from scratch.

If we don't care about consistency, we could streamline this to look a little more like a human wrote it.

It's not quite as simple as *just* downloading the file; we definitely need to visit `prepareRequest`, which does the signing to get the request through AGC's proxy, assuming the URL we got actually points to the server that gave it to us.

I left all the error code detection and error response parsing identical to the WES API request I copied from. If we get an error specified by WES, we'll understand it as specified by WES, and if we get some other error, we won't really understand it but we'll report it anyway. Probably nothing bad would happen if I cut the code, but I took leaving it in as the default.

This should eventually be extended to support S3 or CloudWatch URLs; it could even get AGC away from the `|` task names.

Reply:
I think I'd prefer human-readable code to code that's consistent with the generated code, personally. Generated code is good in that it's generated, but not good for much else 😛

If these URLs being URLs on the WES server itself is common, then I'm fine with continuing to prepare requests and parse errors in the WES style.